jwst/resample/resample_utils.py

```diff
@@ -303,3 +303,28 @@ def decode_context(context, x, y):
         )
 
     return idx
+
+
+def _xy_range_from_bbox(xmin, xmax, ymin, ymax, shape, bbox=None):
+
+    if (xmin is xmax is ymin is ymax) and xmin in (0, None):
+        xmin = ymin = 0
+        ymax, xmax = np.subtract(shape, 1, dtype=int)
+
+        if bbox:
+            ((x1, x2), (y1, y2)) = bbox
+            if x1 > x2:
+                x1, x2 = x2, x1
+            if y1 > y2:
+                y1, y2 = y2, y1
+
+            # use 0.5+/-eps to avoid having edge pixels with flux coming from
+            # a tiny part of input pixels (with low weight). eps can be set
+            # safely to 0.
+            eps = 0.000001
+            xmin = max(0, int(x1 + (0.5 + eps)))
+            ymin = max(0, int(y1 + (0.5 + eps)))
+            xmax = min(xmax, int(x2 + (0.5 - eps)))
+            ymax = min(ymax, int(y2 + (0.5 - eps)))
+
+    return xmin, xmax + 1, ymin, ymax + 1
```

Comment:
Also the logic in this function seems inverted to me. I would think `bbox` should be used by default and overwritten by user values if supplied but it's the opposite - if `bbox` is available it's always used. Is this what is desired?

Reply:
Both `dodrizzle()` in `gwcs_drizzle` and `ResampleData.drizzle_arrays()` in the `resample` module accept `xmin`, `xmax`, `ymin`, `ymax`. This is the only way to indicate what part of the input image should be used. When all `x/y min/max` are 0 => resample the entire input image. That's all. Unfortunately, all these functions are called with these default parameters (`dodrizzle` ultimately is called from `add_image()` using defaults) => resample the entire image.

Looking at https://github.com/spacetelescope/jwst/pull/6041/files#diff-8726cc4bf91b91dd490c35ce61e58c0c840b249b1868fa50cd25a012b13f3d98R248 I see that default values have been changed (but only when resampling variance arrays!) from `0` to `None` without documenting this. Looking at the code, if `x/y min/max` are all `0` then they would just get passed to `cdriz` "as is" (meaning use the entire image) but when `x/y min/max` are all `None`, then the code as implemented in https://github.com/spacetelescope/jwst/pull/6041 would use the bounding box.

IMO this is a bad choice of default parameters (both for `cdriz` and `dodrizzle`, `drizzle_arrays`). I would have preferred `None` to mean "no limits" (use entire image). If there are limits - they should be computed at the top level by the caller, i.e., `add_image`.

The problem that I was trying to fix was to have the same behavior both for resample variance and resample SCI image. Now I realize that because `dodrizzle` has defaults set to 0 (for x and y limits), it is still not using the WCS bounding box as it does when resampling variance arrays.

But before I fix this, can we decide whether it is OK to modify API for these functions (`dodrizzle` and `drizzle_arrays`) to have the same defaults (undo changes to defaults made in  https://github.com/spacetelescope/jwst/pull/6041) and set the limits in the caller? CC: @hbushouse